core/chains/solana/soltxm/txm.go

```diff
@@ -0,0 +1,114 @@
+package soltxm
+
+import (
+	"context"
+	"fmt"
+
+	solanaGo "github.com/gagliardetto/solana-go"
+
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana"
+	solanaClient "github.com/smartcontractkit/chainlink-solana/pkg/solana/client"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/config"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+var (
+	_ services.ServiceCtx = (*Txm)(nil)
+	_ solana.TxManager    = (*Txm)(nil)
+)
+
+// Txm manages transactions for the solana blockchain.
+// simple implementation with no persistently stored txs
+type Txm struct {
+	starter    utils.StartStopOnce
+	lggr       logger.Logger
+	tc         func() (solanaClient.ReaderWriter, error)
+	queue      chan *solanaGo.Transaction
+	stop, done chan struct{}
+	cfg        config.Config
+}
+
+// NewTxm creates a txm. Uses simulation so should only be used to send txes to trusted contracts i.e. OCR.
+func NewTxm(tc func() (solanaClient.ReaderWriter, error), cfg config.Config, lggr logger.Logger) *Txm {
+	lggr = lggr.Named("Txm")
+	return &Txm{
+		starter: utils.StartStopOnce{},
+		tc:      tc,
+		lggr:    lggr,
+		queue:   make(chan *solanaGo.Transaction, 1000), // queue can support 1000 pending txs
+		stop:    make(chan struct{}),
+		done:    make(chan struct{}),
+		cfg:     cfg,
+	}
+}
+
+// Start subscribes to queuing channel and processes them.
+func (txm *Txm) Start(context.Context) error {
+	return txm.starter.StartOnce("solanatxm", func() error {
+		go txm.run()
+		return nil
+	})
+}
+
+func (txm *Txm) run() {
+	defer close(txm.done)
+	ctx, cancel := utils.ContextFromChan(txm.stop)
+	defer cancel()
+	for {
+		select {
+		case tx := <-txm.queue:
+			// TODO: this section could be better optimized for sending TXs quickly
+			// fetch client
+			client, err := txm.tc()
+			if err != nil {
+				txm.lggr.Errorw("failed to get client", "err", err)
+				continue
+			}
+			// process tx
+			sig, err := client.SendTx(ctx, tx)
+			if err != nil {
+				txm.lggr.Criticalw("failed to send transaction", "err", err)
```

Comment:
If it doesn't send is it ever sent again or does it fail forever?

Reply:
at this point, it does not resend which is how the original implementation worked. this is something that is an open todo item - a more robust tx manager with tx confirmation and potentially retries.

right now, it's not super important because of how quickly we submit data to solana. by the time, a tx is noticed to have failed, another OCR round will have passed.